src/main/java/seedu/address/storage/JsonAdaptedPerson.java

```diff
@@ -78,7 +78,7 @@ public Person toModelType(List<Tag> addressBookTagList) throws IllegalValueExcep
         final List<Tag> personTags = new ArrayList<>();
         for (JsonAdaptedTag tag : tagged) {
             for (Tag realTag : addressBookTagList) {
-                if (realTag.isSameTag(tag.toModelType())) {
+                if (realTag.equals(tag.toModelType())) {
                     // Add the unique tag object reference in AddressBook's uniqueTagList to this Person's tags
                     personTags.add(realTag);
                 }
```

Comment:
toModelType with no parameters was just used to avoid a longer line when requiring no tags, it invoking the method shorter (makes tagList an optional argument). The other test cases had an explicit tag list that need to be created, which is why they must use the other method. But do you want to change the last line of toModelType from

`assertThrows(IllegalValueException.class, expectedMessage, person::toModelType);`

to

`assertThrows(IllegalValueException.class, expectedMessage, () -> person.toModelType(new ArrayList<>()));`

Reply:
for the double for loop part, if we remove in equality check and tag doesn't exist in addressBookTagList, won't it return -1 and cause get to throw an IndexOutOfBounds exception

Reply:
How about
```java
        final List<Tag> convertedTags = new ArrayList<>();
        for (JsonAdaptedTag adaptedTag : tagged) {
            convertedTags.add(adaptedTag.toModelType());
        }

        final Set<Tag> modelTags = convertedTags.stream()
                .filter(addressBookTagList::contains)
                .collect(Collectors.toSet());
```
?

Reply:
> for the double for loop part, if we remove in equality check and tag doesn't exist in addressBookTagList, won't it return -1 and cause `get` to throw an IndexOutOfBounds exception
 
Yes. However, it will never be the case that the tag does not exist in addressBookTagList but exists in the current Person.

Assuming the addressbook.json data file has not been wrongly modified, if a tag exists in the person, a tag with the same name **WILL** exist in addressBookTagList.

Reply:
> How about
> 
> ```java
>         final List<Tag> convertedTags = new ArrayList<>();
>         for (JsonAdaptedTag adaptedTag : tagged) {
>             convertedTags.add(adaptedTag.toModelType());
>         }
> 
>         final Set<Tag> modelTags = addressBookTagList.stream()
>                 .filter(convertedTags::contains)
>                 .collect(Collectors.toSet());
> ```
> 
> ?

Yeah, I think this should work

Reply:
Yea sure, there's no issue with the design, you can update the double-for loop with any actually as long as it makes sense during serialisation and deserialisation. (Although I think making addressBookTagList the stream instead of nonUniqueTags the stream can make mapping to the realTags easier?)

Reply:
Uh I should update on my side or do you wanna push commits via remote

Reply:
I think you can update on your side

Reply:
Okay, how about the toModelType(), do we want to make the `List<Tag>` optional or should we get `tagList` to be passed explicitly

Reply:
Sorry I read the toModelType() method wrongly.

But I still think we should make `tagList` be passed explicitly. I think there is a chance toModelType() without parameters could be used wrongly in the future.

For the previous testcases, I created explicit `tagList` to be more consistent. The `VALID_NAME`, `VALID_PHONE`, etc. are from `TypicalPersons.BENSON`, and `TypicalPersons.BENSON` has the tags that I added into the created `tagList` that passed into toModelType().

Reply:
That's a good point.

Everything should be updated now. The loan tests use `BENSON`'s tags now

Reply:
BTW sorry to trouble you, but could you change the rest of the test cases to use your

```java
List<Tag> tagList = new ArrayList<>(BENSON.getTags());
```

instead of the old

```java
List<Tag> tagList = new ArrayList<>();
tagList.add(new Tag("owesMoney"));
tagList.add(new Tag("friends"));
```

I think your style is more readable and less error prone, thanks and sorry for all the trouble!

Reply:
Sure okay, I think there is a difference though, since BENSON's tags actually doesn't include `friends`, I think I'll just create a list field called addressBookTags to include both BENSON's tags + new Tag("friends") and use this list everywhere in this test case

Reply:
Oh no, the `friends` was something I forgot to delete last time. BENSON's tags previously had `friends`, and sometime after updating these tests, I deleted the `friends` tag from BENSON for something else and forgot to delete the `friends` in these testcases.

Reply:
Okay done